### ISSUE TITLE
Falling nodes with node inventory cause crash when unloaded

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#7020]
##### Issue type
 - Bug report
 
##### Minetest version
7b2687ffc681ec927fe960d1e92158a93f925a45

##### OS / Hardware
Operating system: GNU/Linux
CPU: x86_64

##### Summary
If you add metadata to a node, and this node metadata includes an inventory, and this node also is a falling node, and you make this node fall into an unloaded area (forcing it to unload an call `get_staticdata`), the server **crashes** with the following message:

```
  what():  Runtime error from mod '*builtin*' in callback luaentity_GetStaticdata(): ...inetest/git/minetest/bin/../builtin/common/serialize.lua:151: Can't serialize data of type userdata
```

I looked a bit into this. The `userdata` comes from the items in the node's inventory. If the falling node has no inventory, it works fine.

##### Steps to reproduce
* Add `falling_node=1` to `default:chest` from Minetest Game
* Place a chest in a way that it will fall into an unloaded area
* BOOM!

**Comment by paramat** [comment]
Might be solved by #6999

**Comment by paramat** [comment]
Maybe they should halt when just above 'ignore'? Like item entities now do, see #6999

**Comment by Wuzzy2** [comment]
That might be possible. Haven't checked.

**Comment by paramat** [comment]
I might work on it as it's part of my rework of world edge stuff.

**Comment by paramat** [comment]
PR #7024 might fix the issue. From testing, falling node entities do not enter unloaded areas but they do enter ignore (2 different things), i guess your issue is caused by ignore. My PR can delete the entities on contact with either ignore or unloaded area.

**Comment by Warr1024** [comment]
Deleting the falling nodes doesn't really fix the problem, because:

- There are other ways for falling nodes to end up in unloaded areas, and
- We don't necessarily want those nodes to just disappear.

The problem can also be reproduced by using fly/fast/noclip to speed away from a falling node fast enough (or if the node can be triggered to fall with enough delay) that the area containing the falling_node ent tries to unload itself before it can land.  I ran into this issue "in the wild" because I've been working on mods that rely on falling inventories.

Ultimately the issue is that minetest.serialize doesn't know how to serialize *any* kind of userdata, and metaRef:to_table() returns a table which is only *partially* serializable.  This could be fixed by either making metaRef's to/from table functions recurse into the resultant structures and convert all userdata to serializable format, or by teaching minetest.serialize how to detect specific types of userdata and serialize them in some sane way.

#7772 looks related; either of the solutions above would probably produce a workable fix for #7772 as well.

**Comment by Warr1024** [comment]
Problem is reproducible in 0.4.17.  I haven't tested whether it's able to restore the inventory correctly (relying on the fact that *most* places that accept an ItemStack will also accept its string equivelent).  It also only works specifically with falling_node, and not necessarily any other case where a mod may want to serialize node meta refs.

```diff --git a/builtin/game/falling.lua b/builtin/game/falling.lua
index 991962cc..038726bc 100644
--- a/builtin/game/falling.lua
+++ b/builtin/game/falling.lua
@@ -22,7 +22,16 @@ core.register_entity(":__builtin:falling_node", {

        set_node = function(self, node, meta)
                self.node = node
-               self.meta = meta or {}
+               meta = meta or {}
+               if type(meta) ~= "table" then meta = meta:to_table() end
+               for k1, v1 in pairs(meta.inventory or {}) do
+                       for k2, v2 in pairs(v1) do
+                               if type(v2) == "userdata" then
+                                       v1[k2] = v2:to_string()
+                               end
+                       end
+               end
+               self.meta = meta
                self.object:set_properties({
                        is_visible = true,
                        textures = {node.name},
```

The funny thing is that I could swear I once had a mod that relied on being able to serialize an entire mapblock worth of content, including node meta and inventories, to a lua table.  Did NodeMetaRef:to_table() once serialize the ItemStacks to string, but stopped doing this?  Maybe this is a regression...
